opt: define execution operators in optgen

RaduBerinde opened the pull request:
This commit moves the definition of execution operators to an .opt
file. The exec factory (and the stub factory) is now generated code.
This will reduce the amount of manual work for changes and allow some
useful code generation for EXPLAIN.

We extend optgen to allow type strings that include strings, brackets,
etc, so that we can use the Go types. I say we should allow this in
general and avoid the cryptic translations (like `Type` -> `*types.T`).

To avoid the break in uniformity, this change reverts the
`exec.BufferNode` to just an `exec.Node` (this distinction was mostly
for cosmetic reasons).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51899)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Nice!

> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-:-MDFkYcK7lpNDtr_G9f5:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @mjibson, and @RaduBerinde)

___
*[Makefile, line 1612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFgGkc5lYG0BBFWrcT:-MDFgGkc5lYG0BBFWrcU:b-brnlie) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/Makefile#L1612)):*
> ```Makefile
> 	optgen -out $@ factory $(optgen-defs) $(optgen-norm-rules)
> 
> pkg/sql/opt/exec/factory.og.go: $(optgen-defs) $(optgen-norm-rules) bin/optgen
> ```

should `optgen-norm-rules` be `optgen-exec-defs`?
___
*[pkg/sql/opt/exec/factory.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFh20a13wW7CG29YLw:-MDFh20a13wW7CG29YLx:b-gaj11d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/pkg/sql/opt/exec/factory.go#L261)):*
> ```Go
> 	// for a given inserted row). The values passed correspond to InsertCols
> 	// above.
> 	MkErr func(tree.Datums) error
> ```

this can be MkErrFn
___
*[pkg/sql/opt/optgen/cmd/optgen/testdata/execfactory, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFk8TM--qBoN85gYl3:-MDFk8TM--qBoN85gYl4:bff9fs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/pkg/sql/opt/optgen/cmd/optgen/testdata/execfactory#L67)):*
> ```
> 		// columns is a field that has a multi-line
> 		// comment.
> 		columns sqlbase.ResultColumns,
> ```

do we need to import sqlbase?

(looking at the test failures, seems like you might be missing some other imports too)


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-:-MDFt408DhC388_61Z_B:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @mjibson, and @rytaft)

___
*[Makefile, line 1612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFgGkc5lYG0BBFWrcT:-MDFqQB52jI9JVP8tYkr:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/Makefile#L1612)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should `optgen-norm-rules` be `optgen-exec-defs`?
</blockquote></details>

Good catch, fixed.
___
*[pkg/sql/opt/exec/factory.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFh20a13wW7CG29YLw:-MDFt2f375scuZYDT_Rz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/pkg/sql/opt/exec/factory.go#L261)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this can be MkErrFn
</blockquote></details>

Done.
___
*[pkg/sql/opt/optgen/cmd/optgen/testdata/execfactory, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFk8TM--qBoN85gYl3:-MDFqhmm2KlP5XRvqAVU:b-h138zb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/pkg/sql/opt/optgen/cmd/optgen/testdata/execfactory#L67)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

do we need to import sqlbase?

(looking at the test failures, seems like you might be missing some other imports too)
</blockquote></details>

I just forgot to update this test. Fixed.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-:-MDFwKRI2FuJI1pSDMt0:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @mjibson, and @RaduBerinde)

___
*[Makefile, line 1612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFgGkc5lYG0BBFWrcT:-MDFvzJt4CkDthWcYcWO:b-6dp3w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/Makefile#L1612)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good catch, fixed.
</blockquote></details>

rules -> defs?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @mjibson, and @rytaft)

___
*[Makefile, line 1612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFgGkc5lYG0BBFWrcT:-MDFx8qv8QEN0UPne-Z2:bu1vpu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/Makefile#L1612)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

rules -> defs?
</blockquote></details>

:blush: Let's not speak of this again.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, and @mjibson)

___
*[Makefile, line 1612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-MDFgGkc5lYG0BBFWrcT:-MDGN4khD7a3UUOfXgWZ:b-omb9w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc58e867cef5a4fece43803a7f0bba9a2b8506c/Makefile#L1612)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

:blush: Let's not speak of this again.
</blockquote></details>

😆 🤐 lips are sealed


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@andy-kimball let me know if you have any objections to the optgen change to allow go-like types - and if you agree that we should change some of the existing ones so we don't need to remap them.

jordanlewis commented:
Awesome!!

andy-kimball reviewed (commented):
I have no objections to allowing go-like types. I like where you're going here. I do think we're going beyond "optgen" here - if we go much further we should consider elevating this tool to a higher part of the tree and/or renaming it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51899#-:-MDL9mAmDRmlRqKWMVfD:b-ccwiaf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @mgartner, @mjibson, and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/opt/exec/factory.opt:
```diff
@@ -0,0 +1,638 @@
+# Scan returns a node that represents a scan of the given index on
```
It'll be annoying to update, but all these comments say "returns" because they were previously functions. Maybe a quick fix of `%s/returns/represents/` would work?

RaduBerinde replied:
Not sure, the comments do make sense when you jump to the definition of a generated method like `ConstructScan`

RaduBerinde replied:
Maybe if I make the first line of the method comment` ConstructXXX returns a node that implements a XXX execution operator` and then copy the comments that explanin what the `XXX operator` means?

RaduBerinde replied:
Done.

mgartner replied:
I like it 👍

RaduBerinde committed:
opt: define execution operators in optgen

This commit moves the definition of execution operators to an .opt
file. The exec factory (and the stub factory) is now generated code.
This will reduce the amount of manual work for changes and allow some
useful code generation for EXPLAIN.

We extend optgen to allow type strings that include strings, brackets,
etc, so that we can use the Go types. I say we should allow this in
general and avoid the cryptic translations (like `Type` -> `*types.T`).

To avoid the break in uniformity, this change reverts the
`exec.BufferNode` to just an `exec.Node` (this distinction was mostly
for cosmetic reasons).

Release note: None

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r-

RaduBerinde commented:
bors r+
